### ISSUE TITLE
Onboarding the federated learning into multi-cluster environment(ACM)

### ISSUE THREAD

**Issue by yanmxa** [yanmxa/federated-learning#1]
### Containerize the Collaborator and Aggregator Client

Training a neural network often involves third-party packages like torch, which significantly increase the image size. In many cases, the image size can exceed 1GB.

For example:

```bash
299M    /Users/yanmeng/anaconda3/envs/fl/lib/python3.11/site-packages/torch
```

To simplify the initial setup, I used the `sklearn` package to build a lightweight Logistic Regression model for the startup.

Other issues when containerize the application:
- openfl: https://github.com/securefederatedai/openfl/issues/1211
- flower discussion: https://friendly-flower.slack.com/archives/C01RM6LMKQA/p1734316514339399?thread_ts=1734238174.554119&cid=C01RM6LMKQA

### Federated Learning

- Operand: `Namespaced` or `Cluster`
- Samples: https://github.com/yanmxa/federated-learning/tree/main/controller/config/samples
- Different with the [proposal](https://docs.google.com/presentation/d/1i-U2NIlgYiFdV5f4xUVR4r6y1z3ZYVJun7z0kkgOgRo/edit#slide=id.g2f455984551_0_113)
  - Layer configuration is disabled
  - Clients -> Client: Use a single "client" configuration instead of defining individual details for each client. This ensures scalability and flexibility by avoiding repetitive definitions.
  
  - Remove client replicas:
    - Each client should not have replicas. Clients are split based on data distribution, not replication.
    - For a single data source, there should be exactly one client (model). Creating unnecessary replicas wastes training resources and can degrade accuracy during aggregation.
 
   - Use the of placement to schedule the client to specific clusters(`key` of the `Label` or `ClusterClaim` from the managed cluster)

   - Use the `value` of the `Label` or `ClusterClaim` from the managed cluster to locate the data metadata for the client

**Comment by yanmxa** [comment]
TODO:

~1. addon cluster -> fl - client~

2. controller -> manifestwork resplicasset -> cluster namespace

3. namespaced scope -> clusterset, placement

4. result -> model save(s3, kubernetes ...)

**Comment by yanmxa** [comment]
Jan 5, 2025

- The model with more parameters
- Data Privacy -> Case
- Detail the case before the demo

**Comment by yanmxa** [comment]
Case: 

- 10 classes 

cluster1: 5 bird, truck 
cluster2: 5 ...

-> fl  model performance

**Comment by yanmxa** [comment]
Feb 11, 2025

1. Multilayer 

<img width="1052" alt="Image" src="https://github.com/user-attachments/assets/4f7bd913-24d6-4612-a8fc-245a53e2016e" />

- **Does the customer require this design?**  
  Review different Federated Learning frameworks—none of them currently support this kind of architecture.  

- **Gaps:**  
  - **Federated Learning:** Service B needs to function as both an aggregator and collaborator, but FL frameworks don’t support this dual functionality at the moment. This will require integrating these separate features, which may involve delving into the details and implementation of each FL framework.  
  - **OCM:** The **Placement API** needs to be able to schedule workloads on managed clusters. Alternatively, we may need to treat the managed cluster as a hub cluster to support this functionality, similar to the global hub scenarios.

2. Leaf nodes setup  - is node -> hub FL process?
    
<img width="589" alt="Image" src="https://github.com/user-attachments/assets/92542afb-360d-48a6-80ea-b297b316eea0" />

  Note: Currently, we lack a secure connection between the client and server. If we allow external nodes or devices to run the client or collaborator, we’ll need to enable a TLS connection to secure the communication.
    
3. Reference other FL to involve the multicluster environment: https://github.com/yanmxa/federated-learning/issues/3
4. Lack of security connection

5. Local Model Metrics: https://github.com/yanmxa/federated-learning/issues/4

6. Add more framework: https://github.com/open-cluster-management-io/ocm/issues/825

7. Give a interface to use the pretrained model: https://github.com/yanmxa/federated-learning/commit/eeb13ac582f3bce30078c5c03e65dd81cf0bde8f

**Comment by yanmxa** [comment]
give a interface to use the pretrained model

**Comment by yanmxa** [comment]
### Answer 7: Add pre-trained model

 1. Configure the init model
```yaml
cat <<EOF | oc apply -f -
apiVersion: federation-ai.open-cluster-management.io/v1alpha1
kind: FederatedLearning
metadata:
  name: federated-learning-sample
spec:
  framework: flower
  server:
    ...
    storage:
      type: PersistentVolumeClaim
      name: model-pvc
      path: /data/models/model_round_1_2025-02-25-14-27-46.pth
      size: 2Gi
  client:
    ...
EOF
```
2. Verification

<img width="1208" alt="Image" src="https://github.com/user-attachments/assets/2eb39b49-37f7-4fb3-8814-fba63cddab90" />
